---
title: "Alpha diveristy across different ecosystems"
author: "Xin (David) Zhao"
date: "Last edited `r format(Sys.time(), '%d %B %Y')`"
knit: (function(inputFile, encoding) {
      out_dir <- 'docs';
      rmarkdown::render(inputFile,
                        encoding=encoding,
                        output_file=file.path(dirname(inputFile), out_dir, 'index.html'))})
output:
  html_document:
    theme: leonids 
    highlight: pygments
    df_print: paged
    toc: TRUE
    toc_float: TRUE
    collapsed: FALSE
    number_sections: TRUE
    fig_width: 7
    fig_height: 6
    fig_caption: TRUE
editor_options: 
  markdown: 
    wrap: 72
bibliography: references.bib
---

<!-- <html> -->

<!-- <head> -->

<!-- ```{=html} -->
<!-- <style> -->

<!-- h1{ -->
<!--  color: #055C9D; -->
<!--  font-family: Georgia; -->
<!--  font-size: 200% -->
<!-- } -->


<!-- h2{ -->
<!--  color: #055C9D; -->
<!--  font-family: helvetica; -->
<!--  font-size: 150% -->
<!-- } -->

<!-- h3{ -->
<!--  color: #055C9D;   -->
<!--  font-family: helvetica; -->
<!--  font-size: 120%;  -->
<!-- } -->

<!-- p { -->
<!--  color: #333333; -->
<!--  font-family: helvetica; -->
<!--  font-size: 100%; -->
<!-- } -->

<!-- .blackbox { -->
<!--   padding: 1em; -->
<!--   background: green; -->
<!--   color: black; -->
<!--   border: 2px solid orange; -->
<!--   border-radius: 10px; -->
<!-- } -->

<!-- .center { -->
<!--   text-align: center; -->
<!-- } -->

<!-- </style> -->
<!-- ``` -->
<!-- </head> -->

<!-- </html> -->

```{r setup, include = FALSE}
# set options for the entire document 
knitr::opts_chunk$set(fig.align = 'center', 
                      fig.height=6, fig.width=8,
                      dev="png",
                      echo=TRUE, #display code in output document 
                      error=FALSE,
                      collapse = FALSE, 
                      message=FALSE) #stop render when error occurs   
```

## Problem

## Solution

## Project outline

1. Estimate five alpha diversities across major body sites 
2. Compare alpha diversities by body sites + visualization + interpretation 
3. Correlation between various alpha diversities + visualization + interpretation 
4. Multivariate regression on alpha diversities and environmental variables + visualization + interpretation 


## R codes

### Load necessary libraries

```{r library, warning=FALSE, collapse=TRUE}

# load library necessary for this project
# install.package("ggstatsplot", dependencies = TRUE) 
library(ggstatsplot)  # forest plot for model output   

library(tidyverse)   
library(vegan)
library(phyloseq) # import and process qimme output files 
library(phyloseqCompanion)
library(ggpubr)  
library(picante)  # estimate Faith's PD 
library(Hmisc)  # calculate correlation values and p 
library(corrplot) # draw correlation matrix 
library(PerformanceAnalytics) 
library(zCompositions) # treat missing and zero values  
library(caret)  # machine learning package including checking highly correlation variables 
library(lares)  # compute and plot correlations 
library(parameters)  # tabulate model summary 
library(gt) # get a compact and comprehensive summary table in HTML file 


```

### Import metagenomic 16S sequence data and metadata

Human microbiome project (HMP) 16S rRNA gene diversity: examines the
diversity of 16S RNA genes in the human microbiome for the human
microbiome project (HMP)[@aframew2012]

-   Cohort: Center "Healthy Cohort"
-   Data type: 16S metagenomic sequence - NCBI Bioproject ID: 48489
-   300 human subjects, up to 3 time points, 5 body sites
-   Metadata: Unique subject ID, body site, sex and visit number which
    are accessible via [the HMP metadata
    catalog](https://hmpdacc.org/hmp/catalog/grid.php?dataset=metagenomic)

Download Qiime-processed OTU table (v3-5) and phylogenetic trees (v3-5)
from the [website](https://hmpdacc.org/hmp/HMQCP/#data).

```{r import HMP dataset}

# extract OTU table data from .gz files 

otu_v35_hmp <- read.delim(gzfile("C:/Users/17803/Documents/RStuodio-link-GitHub/alpha-diversity-ecosystems/otu_table_v35.txt.gz"),
                      header = FALSE)  # whether the first line describe the column names 

head(otu_v35_hmp)



# unzip the final OTU table after removing mislabeling and contamination 
otu_v35_final_hmp <- read.delim(gzfile("./otu_table_psn_v35.txt.gz"),
                                header = FALSE)  # whether the first line describe the column names 


```


Read in the Qiime-output phylogenetic tree object using `phyloseq` R package. Follow the instruction in the section, `4.5 Import from QIIME Legacy` in the [vignette](https://www.bioconductor.org/packages/release/bioc/vignettes/phyloseq/inst/doc/phyloseq-basics.html#load-phyloseq).



```{r phylogenetic tree, warning=FALSE}

# view the instruction of phyloseq package 
# vignette("phyloseq_analysis")

 
# otu table file 
otu_hmp_phylo <- gzfile("./otu_table_psn_v35.txt.gz") 

# phylogenetic tree file 
ph.tree_hmp_phylo <- gzfile("./rep_set_v35.tre.gz") 


# import both files with phyloseq package 
hmp_phyloseq <- import_qiime(otufilename = otu_hmp_phylo,
                             treefilename = ph.tree_hmp_phylo)

hmp_phyloseq  # phylogenetic object  


```
Now the resulting phyloseq-class object contains an `OTU-table`, `taxomony table` and `phylogenetic tree`.  


Mapping file (16S variable region v35) was accessible from this [website](https://hmpdacc.org/hmp/resources/metagenomics_sequencing_analysis.php).

Convert the imported mapping file to phyloseq class `sample data` using `import_qiime_sample_data()` function of phyloseq package. 


```{r metadata}

sample_data_phylo <- import_qiime_sample_data("./v35_map_uniquebyPSN.txt.bz2")   # import Qiime-output sample data  

```


```{r body site}

table(sample_data_phylo$HMPbodysubsite)   # check the distribution of sub-body sites  

```

Next, merge the sample data and previous phyloseq object into a self-consistent, phyloseq-class object, using `merge_phyloseq`. The resulting phyloseq object contains an OTU table, sample-data, and pylogenetic tree. 


```{r merge phyloseq-class object}

hmp_phylo_data <- merge_phyloseq(sample_data_phylo, hmp_phyloseq)  # merge sample data and previous phyloseq-class object  

```


Calculate alpha diversity by following the [vignettes](https://www.bioconductor.org/packages/release/bioc/vignettes/phyloseq/inst/doc/phyloseq-analysis.html#simple-exploratory-graphics)

```{r filter phyloseq data}

# remove OTU not present in any samples 
hmp_filter <- prune_taxa(taxa_sums(hmp_phylo_data) >0, hmp_phylo_data) 

# alpha diversity metrics of interest 
alpha_metrics <- c("Observed", "Chao1", "ACE", "Shannon", "Simpson", "InvSimpson") 

# plot based on alpha-diversity of stool samples alone 

plot_richness(hmp_filter,  # phyloseq object 
              x = "HMPbodysubsite",  # a variable to map to the x-axis 
              color = "HMPbodysubsite", 
              measures = "Shannon")  + # calculate shannon  
        geom_boxplot(alpha = 0.6)


alpha_plot <- plot_richness(hmp_filter,  # phyloseq object 
                            x = "sex",  # a variable to map to the x-axis 
                            color = "HMPbodysubsite",
                            measures = "Chao1") # calculate chao1 


# plot with ggplot2 
ggplot(data = alpha_plot$data,
       aes(x= HMPbodysubsite, y=value, color = sex), 
       alpha = 0,1) +
        geom_boxplot() +
        coord_flip() 

```

### Estimate alpha diversity metrics using estimate_richness function 

The function, `estimate_richness` in the `phyloseq` package returns several commonly used alpha-diversity metrics:

- Observed
- Chao1
- ACE
- Shannon
- Simpson
- InvSimpson

Next, calculate alpha-diversity metrics across different sub-body sites, referring to this [instruction](https://rpubs.com/lconteville/713954)

```{r alpha-div estimate}

# 18 sub-body sites 
table(sample_data(hmp_filter)$HMPbodysubsite) 

```


Referring to the below map sourced from the human microbiome project, I assigned sub-body sites to five (major) body sites. 

![Where are body sites](Sub-bodysite-cartoon.jpg)


```{r add-a-new-column}

# add body_site to the sample data 
# convert sample data to data frame for following manipulation 
sample_data2 <- phyloseqCompanion::sample.data.frame(hmp_filter) 

class(sample_data2) # check if data frame 

sample_data2 <- sample_data2 %>% 
  mutate(body_site = case_when(
    HMPbodysubsite %in% c("Attached_Keratinized_gingiva",
                          "Buccal_mucosa",
                          "Hard_palate",
                          "Palatine_Tonsils",
                          "Saliva",
                          "Subgingival_plaque",
                          "Supragingival_plaque",
                          "Throat",
                          "Tongue_dorsum") ~ "oral_cavity",
    HMPbodysubsite %in% c(
      "Anterior_nares"
    )~ "nasal_cavtiy",
    HMPbodysubsite %in% c(
      "Left_Antecubital_fossa",
      "Right_Antecubital_fossa",
      "Left_Retroauricular_crease",
      "Right_Retroauricular_crease"
    )~ "skin",
    HMPbodysubsite %in% c(
      "Stool"
    ) ~ "gastrointestinal_tract",
    HMPbodysubsite %in% c(
      "Mid_vagina",
      "Posterior_fornix",
      "Vaginal_introitus"
    ) ~ "urogenital_tract")
    )

# check the new column 
table(sample_data2$body_site)


# integrate the updated sample data into the phyloseq object 

sample_data3 <- sample_data(sample_data2)

hmp_filter2 <- merge_phyloseq(sample_data3, hmp_phyloseq)  


```


Estimate alpha-diversity measures by body sites 

```{r alpha-diversity-5-body-sites, warning=FALSE}

alpha_div_hmp <- phyloseq::estimate_richness(hmp_filter2, 
                                             split = TRUE, 
                                             measures = c("Observed", # remove Fisher since vegan implementation dose not work anymore  
                                                          "Chao1", 
                                                          "ACE", 
                                                          "Shannon", 
                                                          "Simpson", 
                                                          "InvSimpson"))

head(alpha_div_hmp) # view first rows 

```

Next, compare various alpha-diversity metrics across five body sites - Are alpha diversity different among body sites? 

```{r}

# check if normal distribution for Shannon 

hist(alpha_div_hmp$Shannon, main = "Shannon index", xlab = "") 


```


It seems somewhat right-shewed. Run anova test to evaluate whether body sites are associated with Shannon diversity. 

```{r shannon-anova}

anova.sh <- aov(alpha_div_hmp$Shannon ~ sample_data(hmp_filter2)$body_site)  
summary(anova.sh)

```
Run a post hoc test, Tukey honest significance difference (Tukey's HSD) after the ANOVA test. 

```{r post-hoc-test}

TukeyHSD(anova.sh)

```


Alternatively, run Kruskal-Wallis rank sum test on non-normal distributed dataset. 

```{r kw-test}

kruskal.test(alpha_div_hmp$Shannon ~ sample_data(hmp_filter2)$body_site) 

```
Get p-values of the Wilcoxon tests comparing each pair of groups. 

```{r wilcoxon-test}

pairwise.wilcox.test(alpha_div_hmp$Shannon, sample_data(hmp_filter2)$body_site, 
                     p.adj = "bonf")  # multiple comparison adjustment - bonferroni 


```


### Estimate Faith's PD using picante R package 

The package `picante` can estimate the phylogenetic aware metric, Faith's PD, which `phyloseq` package cannot do. 

Apply `pd` function to OTU table and phylogenetic tree, following the [instruction](https://rdrr.io/rforge/picante/man/pd.html). The function `pd` returns a data frame of the `PD` and species (`SR`) values for all samples. 

```{r faith PD, warning=FALSE}



# faith_pd <- picante::pd(samp = t(otu_table(hmp_filter2)), 
#                         tree = phy_tree(hmp_filter2), 
#                         include.root = F)  # tree need not rooted, but in case of single-species samples the PD will be equal to NA 

# head(faith_pd) 

# As picante::pd() takes a long, long time to run, I export faith_pd as an external file for generating html. 


# write.csv(faith_pd,
#           file = "faith_pd_HMP.csv") 

faith_pd <- read.csv("faith_pd_HMP.csv", header = T,
                     col.names = c("SampleID", "PD", "SR")) 


```

### Tabulate descriptive statsitics of five alpha diversity metrics 

```{r tabulate-stats}

str(faith_pd)  # PD data frame 

str(alpha_div_hmp)  # other alpha diversities 

str(sample_data2)  # metadata 

```


```{r merge-dataframes, warning=FALSE}

# prepare faith's PD table 
faith_pd_sub <- faith_pd %>% 
  # rownames_to_column("SampleID") %>% 
  dplyr::select(-SR)

faith_pd_sub$SampleID <- as.character(faith_pd_sub$SampleID)


# prepare pyloseq-output alpha diversities 
alpha_div_hmp_sub <- alpha_div_hmp %>% 
  rownames_to_column("SampleID") %>% 
  dplyr::mutate(SampleID2 = str_sub(SampleID, 2)) %>% 
  dplyr::select(-SampleID) %>% 
  rename(SampleID = SampleID2) %>% 
  dplyr::select(SampleID, Chao1, se.chao1, Shannon, Simpson, InvSimpson)


# prepare metadata 
sample_data_sub <- sample_data2 %>% 
  dplyr::select(SampleID = X.SampleID, body_site) 

sample_data_sub$SampleID <- as.character(sample_data_sub$SampleID)  # convert to character variable 


# merge three the above tables by Sample ID using full merge

merg_alpha <- sample_data_sub %>% 
  dplyr::full_join(alpha_div_hmp_sub, by= "SampleID") %>% 
  dplyr::full_join(faith_pd_sub, by= "SampleID")

head(merg_alpha) # view first rows 


```

Calculate summary statistics for five alpha diversities with body site groups 

```{r summary-stats}

summ_alpha <- merg_alpha %>% 
  group_by(body_site) %>% 
  dplyr::summarize(chao_ave = mean(Chao1),  # average for Chao1 
                   chao_std = sd(Chao1), # standard deviations for Chao1  
                   shannon_ave = mean(Shannon),  # Shannon average 
                   shannon_std = sd(Shannon),  # Shannon std 
                   simpson_ave = mean(Simpson),  # Simpson average 
                   simpson_std = sd(Simpson),  # Simpson std 
                   invsimp_ave = mean(InvSimpson),  # InvSimpson average 
                   invsimp_std = sd(InvSimpson), # InvSimpson std 
                   pd_ave = mean(PD),  # PD average 
                   pd_std = sd(PD),  # PD std 
                   sub_size = n())  # group sizes 

summ_alpha_dig <- summ_alpha 

summ_alpha_dig[,-1] <- apply(summ_alpha_dig[,-1], 2, function(x) round(x, 2))  # two decimal digits 

summ_alpha_dig # print the resulting table  
  
```

Draw bar plots with error bars to illustrate the summary stats. 

Refer to the instruction [here](https://stringr.tidyverse.org/reference/str_extract.html) to set regular expression for `str_extract` function. 

```{r barplot-error}

# prepare data 
summ_alpha_dig2 <- summ_alpha_dig %>% 
  tidyr::gather(key = "alpha_metrics", 
                value = "value", 
                -c("body_site", 
                   "sub_size")) %>%
  dplyr::mutate(stats_metrics = str_sub(alpha_metrics, -3)) %>% 
  dplyr::mutate(alpha_metric = str_extract(alpha_metrics, "[a-z]+"))  %>% 
  dplyr::select(-alpha_metrics) %>% 
  tidyr::spread(key = "stats_metrics", value = "value") 
  

summ_alpha_dig2  # print the table 

```

```{r barplot-error-bar, warning=FALSE}

summ_alpha_dig3 <- summ_alpha_dig2 %>% 
  mutate(Ymax = ave + std) %>% 
  mutate(Ymin = ave - std)

ggplot(summ_alpha_dig3, 
       aes(x = alpha_metric, y = ave, fill = body_site)) + 
  geom_col(position = "dodge") +
  geom_errorbar(data = summ_alpha_dig3,
                position = "dodge",
                aes(ymax= Ymax, 
                ymin = Ymin),
                na.rm = T,
                size = 0.1,
                linetype =1,
                color = "gray60") +  
  facet_wrap(vars(alpha_metric),
             scales = "free") 
  


```


### Correlaion matrix between five alpha diveristies 

Use `rcorr` function in `Hmisc` package following the instruction [here](http://www.sthda.com/english/wiki/correlation-matrix-a-quick-start-guide-to-analyze-format-and-visualize-a-correlation-matrix-using-r-software). 


```{r cor-analysis}

head(merg_alpha)  # original data frame that stores all alpha diversities  

```


```{r, warning=FALSE} 

chart.Correlation(merg_alpha[,-c(1,2,4)], histogram=TRUE, pch=19) 
 
```

```{r}
# prepare matrix 

merg_alpha_corr_mat <- merg_alpha %>% 
  dplyr::select(-body_site, -se.chao1, -SampleID)  %>% 
  as.matrix()

rownames(merg_alpha_corr_mat) <- merg_alpha$SampleID


# use Hmisc package to implement correlation analysis 
corr_alpha <- rcorr(merg_alpha_corr_mat, type = "spearman")   

corr_alpha 


```

```{r extract p and rho} 

# ++++++++++++++++++++++++++++
# flattenCorrMatrix
# ++++++++++++++++++++++++++++
# cormat : matrix of the correlation coefficients
# pmat : matrix of the correlation p-values
flattenCorrMatrix <- function(cormat, pmat) {
  ut <- upper.tri(cormat)
  data.frame(
    row = rownames(cormat)[row(cormat)[ut]],
    column = rownames(cormat)[col(cormat)[ut]],
    cor  =(cormat)[ut],
    p = pmat[ut]
    )
}

```


```{r save-data-frame}

flattenCorrMatrix(corr_alpha$r, corr_alpha$P)  

```

Plot the Spearman-based correlation matrix using `corrplot` function in `corrplot` package following the online [guide](http://www.sthda.com/english/wiki/correlation-matrix-a-quick-start-guide-to-analyze-format-and-visualize-a-correlation-matrix-using-r-software). 

```{r plot corr, collapse=TRUE, warning=FALSE}

# Insignificant correlations are leaved blank

corrplot(corr_alpha$r,
         type="upper",
         order="original")  
         # p.mat = p_corr,
         # sig.level = 0.05,
         # insig = "blank") 

```

The correlation matrix plot illustrate that all five alpha diversity metrics are significantly positively associated with each other (p < 0.05), yet with various strength (according to rho). In particular, non-phylogenetic metric, Chao1 and phylogenetic Faith's PD is the most closely correlated pair among others; Shannon is the second most closely associated with Faith's PD. 



### Multivariate regression between the alpha diversity and environmental variables 

Use MasLin2 package to evaluate impact of specific environmental variables on the richness or evenness of bacterial communities.  

The covariables present in the metadata includes `visitno`, `sex`, `RUNCENTER`, and `body_site`. 

```{r metadata covariable}

dplyr::glimpse(sample_data2) # structure of the metadata 

table(sample_data2$RUNCENTER)  # run center distribution 

```

Check whether zero values/ missing values are present in covariables. 

Treat zeroness with proper method, with referring to the `Chap 10.1.4.3` "How to deal zeros in compositional data analysis" in the Book, Statistical Analysis of Microbiome Data with R. Try to use `zComposition` package to treat zero values in categorical variables. 


```{r check zero, collapse=TRUE}

# subset covariables of interest 
sample_data2_cov <- sample_data2  %>%
  dplyr::select(SampleID = X.SampleID, visitno, sex, RUNCENTER, body_site) 

# apply zPatterns function to find and display missing and zero values 
# zPatterns(sample_data2_cov,
#           label = NA, 
#           plot= TRUE)       

```

The result of the function `zPatterns` indicated that there is no missing and zero values in the four covariables and sample ID. 




Check if the four covariables are highly correlated. As all covariables are categorical, I apply the [chi-squared test](https://en.wikipedia.org/wiki/Chi-squared_test) to evaluate in the contingency tables of covariables when sample size large. 

Null hypothesis is the row and the column variables of the contingency table are independent; alternative hypothesis is row and column variables are dependent. Note that the assumption of chi-squared test is that the expected frequency of any cell is at least 5. 

Refer to [sthda.com](http://www.sthda.com/english/wiki/chi-square-test-of-independence-in-r) for R coding to implement chi-squared test.


```{r highly-corr}

# convert covariables of interest to categorical 

table(sample_data2_cov$visitno)  # contingency table for visit no - three levels 1, 2 and 3

table(sample_data2_cov$sex) # two levels female and male 

table(sample_data2_cov$RUNCENTER)  # four levels - BCM, BI, JCVI, and WUGC 

# accumulate sub-groups into a single major group 
sample_data2_cov <- sample_data2_cov %>% 
  mutate(runcenter = case_when(RUNCENTER %in% c("BCM", "BCM,BI", "BCM,JCVI", "BCM,WUGC")~"BCM",
                               RUNCENTER %in% c("BI", "BI,BCM")~"BI",
                               RUNCENTER %in% c("JCVI", "JCVI,BI", "JCBI,WUGC")~"JCVI",
                               RUNCENTER %in% c("WUGC", "WUGC,BCM","WUGC,JCVI")~"WUGC"))

table(sample_data2_cov$body_site) # five levels - gastrointestinal tract, nasal cavity, oral cavity, skin and urogenital tract 



# convert to factor variables 
sample_data2_cov2 <- sample_data2_cov %>% 
  dplyr::select(-SampleID) %>% 
  dplyr::select(-RUNCENTER) %>% 
  dplyr::mutate(visitno = factor(visitno)) %>% 
  mutate(sex = factor(sex, levels = c("female", "male"))) %>% 
  mutate(body_site = factor(body_site)) %>% 
  mutate(runcenter = factor(runcenter))
```


Apply `lares` package to compute and plot correlations between categorical variables. View this [web site](https://statsandr.com/blog/correlogram-in-r-how-to-highlight-the-most-correlated-variables-in-a-dataset/) for R codes and plotting examples. 

```{r chisquared test, warning=FALSE} 

lares::corr_cross(sample_data2_cov2,  # name of dataset 
                  max_pvalue = 0.05,  # display only significant correlation (at 5% level)
                  top = 10) # display top 10 couples of variables (by correlation coefficient) 

```

The sex (`sex`) and urogenital tract (`body_site`) are correlated with low-to-medium strength while other pair-wise correlations are low-to-weak. 



Build multiple linear regression between alpha diversities and four covariables.   

```{r linear-regression}

# alpha diversities data frame 
head(merg_alpha_corr_mat)


# merge covariables and the alpha diversities 
# alpha diversity data frame 
merg_alpha_corr_df <- merg_alpha_corr_mat %>% 
  as.data.frame() %>% 
  rownames_to_column("SampleID") 

dim(merg_alpha_corr_df)  # 4743 observations 

```


Check missing/zero values in dependent variables using `zCompositions` package. 

```{r check NA in depedent variables}

zPatterns(merg_alpha_corr_df,
          label = NA, 
          plot= TRUE)   

```
Missing values are present in one dependent variable, `PD`. 


Check visually whether dependent variables normally distribute. 

```{r normal distribution}

merg_alpha_corr_df %>% 
  gather(key = "alpha_metrics", value = "values", -SampleID) %>% 
  ggplot(aes(x = values, y =..density..))+
  geom_histogram(stat = "bin", binwidth = 1, alpha = 0.5) +
  facet_wrap(vars(alpha_metrics),
             scales = "free") +
  geom_density(color = "blue",
               linewidth = 1,
               alpha = 0.1)

``` 

All five alpha diversities seem shewed to various degree. Use data transformation methods to generate normally distributed dependent variables. The website [GUSTA ME](https://sites.google.com/site/mb3gustame/reference/transformations) provides a collection of common data transformation methods, including logarithmic, square root, power transformations, and box-cox. 


```{r}

# covariables data frame 
sample_data2_cov2_df <- sample_data2_cov2 %>% 
  rownames_to_column("SampleID") 

dim(sample_data2_cov2_df)  # 4743 observations 


# join two data frames by sample ID  
alpha_cov_df <- merg_alpha_corr_df %>% 
  full_join(sample_data2_cov2_df, by= "SampleID")  

dim(alpha_cov_df)  # 4743 observations 
  
```


```{r variable names}

names(alpha_cov_df) 

```

Dependent variables: 
`Chao1`
`Shannon`
`Simpson`
`InvSimpson`
`PD`

Independent variables: 
`visitno`
`sex`
`body_site` 
`runcenter`

```{r linear mod}

# Chao1 model 
mod_chao1 <- lm(Chao1 ~  body_site + visitno + sex + runcenter,
   data = alpha_cov_df,
   na.action =  na.omit # omit NA 
   )

# summary(mod_chao1)  # coefficients and significance of the resulting model 



#  Shannon mod 
mod_shann <- lm(Shannon ~  body_site + visitno + sex + runcenter,
   data = alpha_cov_df,
   na.action =  na.omit # omit NA 
   ) 


# Simpson mod 
mod_simpson <- lm(Simpson ~  body_site + visitno + sex + runcenter,
   data = alpha_cov_df,
   na.action =  na.omit # omit NA 
   ) 


# Inverse Simpson mod 
mod_invsimpson <- lm(InvSimpson ~  body_site + visitno + sex + runcenter,
   data = alpha_cov_df,
   na.action =  na.omit # omit NA 
   ) 


# PD mod 
mod_pd <- lm(PD ~  body_site + visitno + sex + runcenter,
   data = alpha_cov_df,
   na.action =  na.omit # omit NA 
   ) 


```

The model coefficients tell us that, 
- Body sites are associated with alpha diversity, Chao1. Oral cavity has highest diversity compared to other niches. 
- Visit number (sampling time) is linked to alpha diversity, Chao1.  
- Run centers are associated with alpha diversity, Chao1, suggesting that technical variations exist in the data set. 
- Sex is linked to alpha diversity, Chao1, male has higher alpha diversity than female does. 

Overall, four covariables collectively explain 47.35% variation of average alpha diversity, Chao1. 


#### Visualize and tabulate model output 

The post in the blog [Stats and R](https://statsandr.com/blog/multiple-linear-regression-made-simple/#coefficient-of-determination-r2) provides a few R packages to visualize or tabulate linear models, including, 

- `ggstatsplot` 
- `ggstance` 
- `parameters` (tabulate)
- `gt` (tabulate) 
- `equationmatic` (extract model's equation) 
- `report` (automatic reporting) 

In the report, I use `ggstatsplot` to visualize while use `gt` to tabulate the model output for illustration purpose. 

```{r viz model}

# forest plot to show model output in detailed
ggstatsplot::ggcoefstats(mod_chao1)   

```

In this plot:

- when the solid line does not cross the vertical dashed line, the coefficient is different from 0 at the 5% significance level
- a point to the right or to the left of the vertical dashed line indicates positive or negative associations between two variables 
- the more extreme the point, the stronger the relationship 


Next, tabulate the model output in a HTML-friendly manner using the packages `gt` and `parameters`

```{r tabulate mode output}

print_html(model_parameters(mod_chao1, summary = TRUE))  



```

#### Predict Chao 1 from four variables 


```{r pred chao1}


names(alpha_cov_df)


# create a new data 
new_df <- alpha_cov_df[1:3,7:10] 


# predict new individuals with 95% prediction interval 
predict(mod_chao1,
        new = new_df,
        interval = "prediction",
        level = .95) 


```


### Diagnose the models 

Check up whether regression assumption hold true. The website [sthda](http://www.sthda.com/english/articles/39-regression-model-diagnostics/161-linear-regression-assumptions-and-diagnostics-in-r-essentials/#regression-assumptions) provides detailed instructions on diagnosing models. 

Regression assumptions:
- linearity of the data 
- normality of residuals 
- homogeneity of residuals variance 
- independence of residuals error terms 

All these assumptions and potential problems can be checked by producing some diagnostic plots visualizing the residual errors. 


```{r}

mod_chao1_diag_metrics <- broom::augment(mod_chao1)

head(mod_chao1_diag_metrics)

```


```{r dignose mod}

# diagnose plots for chao1 model 
par(mfrow = c(2,2))

plot(mod_chao1)


```


Look more closely at the Cook's distance to identify any influencer points. 

```{r cook distance}

plot(mod_chao1, 4,
     id.n=5) # lable top most extreme values 

plot(mod_chao1, 5)

```


```{r }

# look further the top extreme values 
mod_chao1_diag_metrics %>% 
  top_n(3, wt= .cooksd)

```













