---
title: "Alpha diveristy across different ecosystems"
author: "Xin (David) Zhao"
date: "Last edited `r format(Sys.time(), '%d %B %Y')`"
knit: (function(inputFile, encoding) {
      out_dir <- 'docs';
      rmarkdown::render(inputFile,
                        encoding=encoding,
                        output_file=file.path(dirname(inputFile), out_dir, 'index.html'))})
output:
  html_document:
    # theme: cosmo
    highlight: pygments
    df_print: paged
    toc: TRUE
    toc_float: TRUE
    collapsed: FALSE
    number_sections: TRUE
    fig_width: 7
    fig_height: 6
    fig_caption: TRUE
editor_options: 
  markdown: 
    wrap: 72
bibliography: references.bib
---

<html>

<head>

```{=html}
<style>

h1{
 color: #055C9D;
 font-family: Georgia;
 font-size: 200%
}


h2{
 color: #055C9D;
 font-family: helvetica;
 font-size: 150%
}

h3{
 color: #055C9D;  
 font-family: helvetica;
 font-size: 120%; 
}

p {
 color: #333333;
 font-family: helvetica;
 font-size: 100%;
}

.blackbox {
  padding: 1em;
  background: green;
  color: black;
  border: 2px solid orange;
  border-radius: 10px;
}

.center {
  text-align: center;
}

</style>
```
</head>

</html>

```{r setup, include = FALSE}
# set options for the entire document 
knitr::opts_chunk$set(fig.align = 'center', 
                      fig.height=6, fig.width=8,
                      dev="png",
                      echo=TRUE, #display code in output document 
                      error=FALSE,
                      collapse = FALSE, 
                      message=FALSE) #stop render when error occurs   
```

## Problem

## Solution

## Project outline

## R codes

### Load necessary libraries

```{r library, warning=FALSE, collapse=TRUE}

# load library necessary for this project

library(tidyverse) 
library(vegan)
library(phyloseq) # import and process qimme output files 


```

### Import metagenomic 16S sequence data and metadata

Human microbiome project (HMP) 16S rRNA gene diversity: examines the
diversity of 16S RNA genes in the human microbiome for the human
microbiome project (HMP)[@aframew2012]

-   Cohort: Center "Healthy Cohort"
-   Data type: 16S metagenomic sequence - NCBI Bioproject ID: 48489
-   300 human subjects, up to 3 time points, 5 body sites
-   Metadata: Unique subject ID, body site, sex and visit number which
    are accessible via [the HMP metadata
    catalog](https://hmpdacc.org/hmp/catalog/grid.php?dataset=metagenomic)

Download Qiime-processed OTU table (v3-5) and phylogenetic trees (v3-5)
from the [website](https://hmpdacc.org/hmp/HMQCP/#data).

```{r import HMP dataset}

# extract OTU table data from .gz files 

otu_v35_hmp <- read.delim(gzfile("C:/Users/17803/Documents/RStuodio-link-GitHub/alpha-diversity-ecosystems/otu_table_v35.txt.gz"),
                      header = FALSE)  # whether the first line describe the column names 

head(otu_v35_hmp)



# unzip the final OTU table after removing mislabeling and contamination 
otu_v35_final_hmp <- read.delim(gzfile("./otu_table_psn_v35.txt.gz"),
                                header = FALSE)  # whether the first line describe the column names 


head(otu_v35_final_hmp[,ncol(otu_v35_final_hmp)]) 

```


Read in the Qiime-output phylogenetic tree object using `phyloseq` R package. Follow the instruction in the section, `4.5 Import from QIIME Legacy` in the [vignette](https://www.bioconductor.org/packages/release/bioc/vignettes/phyloseq/inst/doc/phyloseq-basics.html#load-phyloseq).




```{r phylogenetic tree}

# view the instruction of phyloseq package 
# vignette("phyloseq_analysis")

 
# otu table file 
otu_hmp_phylo <- gzfile("./otu_table_psn_v35.txt.gz") 

# phylogenetic tree file 
ph.tree_hmp_phylo <- gzfile("./rep_set_v35.tre.gz") 


# import both files with phyloseq package 
hmp_phyloseq <- import_qiime(otufilename = otu_hmp_phylo,
                             treefilename = ph.tree_hmp_phylo)

hmp_phyloseq  # phylogenetic object 


```
Now the resulting phyloseq-class object contains an `OTU-table`, `taxomony table` and `phylogenetic tree`.  


Mapping file (16S variable region v35) was accessible from this [website](https://hmpdacc.org/hmp/resources/metagenomics_sequencing_analysis.php).

Convert the imported mapping file to phyloseq class `sample data` using `import_qiime_sample_data()` function of phyloseq package. 


```{r metadata}

sample_data_phylo <- import_qiime_sample_data("./v35_map_uniquebyPSN.txt.bz2")   # import Qiime-output sample data  

```


```{r body site}

table(sample_data_phylo$HMPbodysubsite)   # check the distribution of sub-body sites  

```

Next, merge the sample data and previous phyloseq object into a self-consistent, phyloseq-class object, using `merge_phyloseq`. The resulting phyloseq object contains an OTU table, sample-data, and pylogenetic tree. 


```{r merge phyloseq-class object}

hmp_phylo_data <- merge_phyloseq(sample_data_phylo, hmp_phyloseq)  # merge sample data and previous phyloseq-class object  


```


Calculate alpha diversity by following the [vignettes](https://www.bioconductor.org/packages/release/bioc/vignettes/phyloseq/inst/doc/phyloseq-analysis.html#simple-exploratory-graphics)

```{r filter phyloseq data}

# remove OTU not present in any samples 
hmp_filter <- prune_taxa(taxa_sums(hmp_phylo_data) >0, hmp_phylo_data) 

# alpha diversity metrics of interest 
alpha_metrics <- c("Observed", "Chao1", "ACE", "Shannon", "Simpson", "InvSimpson") 

# plot based on alpha-diversity of stool samples alone 

plot_richness(hmp_filter,  # phyloseq object 
              x = "HMPbodysubsite",  # a variable to map to the x-axis 
              color = "HMPbodysubsite", 
              measures = "Shannon")  + # calculate shannon  
        geom_boxplot(alpha = 0.6)




alpha_plot <- plot_richness(hmp_filter,  # phyloseq object 
                            x = "sex",  # a variable to map to the x-axis 
                            color = "HMPbodysubsite",
                            measures = "Chao1") # calculate chao1 


# plot with ggplot2 
ggplot(data = alpha_plot$data,
       aes(x= HMPbodysubsite, y=value, color = sex), 
       alpha = 0,1) +
        geom_boxplot() +
        coord_flip()

```






